torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
         return_fn,
         fqn_to_param_info,
     )
+
+
+def _gather_orig_param_state(
+    fqn: str,
+    fsdp_param_info: FSDPParamInfo,
+    optim_state: Dict[str, Any],
+    shard_state: bool,
+) -> Dict[str, Any]:
+    """
+    Gather the optimizer state for the original parameter with the name ``fqn``.
+    This API should only be used when ``use_orig_params`` is True.
+    """
+    fsdp_state = fsdp_param_info.state
+    flat_param = fsdp_param_info.flat_param
+    param_idx = fsdp_param_info.param_indices[fqn]
+    if (
+        fsdp_state.world_size == 1
+        or fsdp_state.sharding_strategy == ShardingStrategy.NO_SHARD
+    ):
+        return optim_state
+
+    # Gathering state from all ranks. This step may be slow. However,
+    # `state_dict()` is not in the critical path. We can fuse the communication
+    # if the performance becomes a problem.
+    state_objects = {
+        state_name: value for state_name, value in sorted_items(optim_state)
+    }
+    object_list: List[Dict[str, Any]] = [
+        {} for _ in range(cast(int, fsdp_state.world_size))
+    ]
+    dist.all_gather_object(object_list, state_objects)
```

Comment:
I forget; for the `use_orig_params=False` case, what collective do we use? `all_gather_into_tensor`?

Reply:
That's correct.